src/sched.rs

```diff
@@ -0,0 +1,79 @@
+//! APIs for interacting with the scheduler and with processes,
+//! corresponding to <linux/sched.h> and related header files.
+#![allow(improper_ctypes)]
+
+use alloc::vec::Vec;
+use core::ptr;
+
+use crate::bindings;
+use crate::rcu;
+
+extern "C" {
+    fn task_lock_helper(p: *mut bindings::task_struct);
+    fn task_unlock_helper(p: *mut bindings::task_struct);
+    fn next_task_helper(p: *mut bindings::task_struct) -> *mut bindings::task_struct;
+}
+
+/// Represents a `struct task_struct *`.
+pub struct TaskStruct<'a>(&'a mut bindings::task_struct);
+
+impl TaskStruct<'_> {
+    /// Returns the command name / process title. This is a short name,
+    /// typically the base name of the command, and does not have the
+    /// full path or arguments. It's a fixed-sized set of bytes, but by
+    /// convention it's interpreted as NUL-terminated.
+    pub fn comm(&mut self) -> Vec<u8> {
```

Comment:
No practical way to return an `&[u8]`?

Reply:
`self.0.comm` is only valid while we've got a lock on `self.0`, and I don't want to hold a lock for the lifetime of this structure.

We could probably expose the locking to Rust, i.e, something like
```rust
impl TaskStruct {
    fn lock(&mut self) -> LockedTaskStruct;
}
impl LockedTaskStruct {
    fn comm(&mut self) -> &[u8];
}
```
or perhaps even something like
```rust
impl TaskStruct {
    fn comm(&mut self, g: &mut LockGuard) -> &[u8];
}
```
though given the lack of NLL, I'm more hesitant about using an RAII guard to lock a task_struct than I am about making an RCU read-side critical section. An RcuReadGuard only prevents cleanup after making changes, it doesn't prevent a writer from doing the reading, copying, or updating.

Reply:
Is there a reason not to return a `[u8; TASK_COMM_LEN]`? This is basically what [`get_task_comm`](https://github.com/torvalds/linux/blob/v5.8/include/linux/sched.h#L1733-L1736) does in C.

Reply:
Hm, that would work. I was actually leaning towards taking an `&mut [u8]` but I guess yeah you can just return the array and make the compiler deal with it.

(This code was largely just "make anything work at all")

Reply:
Switched it to returning a `[u8; TASK_COMM_LEN]`.

It seems that many of my task structs have some garbage after the null byte. For instance, if in Python you `open("/proc/self/comm").write("foo")`, you'll find that your comm is now `foo\0on\0\0\0...` because it was previously `python`. I could return this directly, but because we support memory safety here on Team Rust, and because I had to do shenanigans to convert i8 -> u8 anyway, I opted to cut it off after the first NUL byte.